File: scripts/moveonjoy_updater.py
#!/usr/bin/env python3
"""
MoveOnJoy Updater v4.1 — Per-Channel Smart Failover + Auto-Restore (Option A)
- Place in: scripts/moveonjoy_updater.py
- Playlist expected: PrimeVision/us.m3u (auto-detected)
- Requirements: requests
"""

from pathlib import Path
import re
import time
import subprocess
from datetime import datetime
import requests
import sys

# ---------- CONFIG ----------
PLAYLIST_CANDIDATES = [
    Path("PrimeVision/us.m3u"),
    Path("IPTV-LIVE/PrimeVision/us.m3u"),
    Path("gtvservices5/IPTV-LIVE/PrimeVision/us.m3u"),
    Path("us.m3u"),
]

SUB_MIN = 3
SUB_MAX = 50
SUB_RANGE = list(range(SUB_MAX, SUB_MIN - 1, -1))  # fl50 -> fl3

RETRIES = 3
RETRY_DELAY = 1  # seconds
SAMPLE_LIMIT = 10  # when sampling lines to test a domain
COOLDOWN_SECONDS = 3600  # 1 hour between commits (set 0 to disable)
LAST_UPDATE_FILE = Path(".moveonjoy_last_update")

# Paths we treat as "special" (prefer to keep on their own working subdomain)
SPECIAL_CHANNEL_PATHS = {
    "DISNEY/index.m3u8",
    # add other exact relative paths here if needed
}

# Regex to parse MoveOnJoy lines like: https://fl25.moveonjoy.com/PATH/index.m3u8
FL_RE = re.compile(r"https://(fl\d+)\.moveonjoy\.com/([^\s]+)")

# ---------- UTILITIES ----------

def log(*args, **kwargs):
    ts = datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S UTC")
    print(f"[{ts}]", *args, **kwargs)

def find_playlist():
    """Return Path to us.m3u or None"""
    for p in PLAYLIST_CANDIDATES:
        if p.exists():
            log("Found playlist at:", p)
            return p
    # recursive search fallback
    for p in Path(".").rglob("us.m3u"):
        log("Found playlist via recursive search:", p)
        return p
    return None

def http_get(url, timeout=6, stream=False):
    for attempt in range(1, RETRIES + 1):
        try:
            r = requests.get(url, timeout=timeout, stream=stream)
            return r
        except requests.RequestException as e:
            if attempt < RETRIES:
                time.sleep(RETRY_DELAY)
            else:
                return None

def check_playlist_has_segments(text):
    """Basic heuristic if m3u8 contains segments or variant playlists."""
    if not text:
        return False
    for L in text.splitlines():
        L = L.strip()
        if L and not L.startswith("#") and (L.endswith(".ts") or L.endswith(".m3u8")):
            return True
    return False

def check_stream_playable(m3u8_url):
    """Return True if playlist reachable and the first referenced segment is reachable."""
    r = http_get(m3u8_url, timeout=6)
    if not r or r.status_code >= 400:
        return False
    text = r.text
    if not check_playlist_has_segments(text):
        # some servers don't list segments; treat as not playable
        return False
    # find first candidate segment or nested m3u8
    for L in text.splitlines():
        L = L.strip()
        if not L or L.startswith("#"):
            continue
        # candidate
        if L.startswith("http"):
            seg_url = L
        else:
            seg_url = m3u8_url.rsplit("/", 1)[0] + "/" + L
        seg_r = http_get(seg_url, timeout=6)
        return (seg_r is not None) and (seg_r.status_code < 400)
    return False

# ---------- PLAYLIST HELPERS ----------

def read_playlist_lines(m3u_path):
    text = m3u_path.read_text(encoding="utf-8")
    lines = text.splitlines()
    return lines, text

def write_playlist_lines(m3u_path, new_lines):
    txt = "\n".join(new_lines)
    # ensure final newline
    if not txt.endswith("\n"):
        txt += "\n"
    m3u_path.write_text(txt, encoding="utf-8")

def enumerate_lines_using_subdomain(lines, subdomain):
    """Yield (idx, full_url, relative_path) lines that reference this subdomain."""
    for idx, line in enumerate(lines):
        if "http" in line and f"{subdomain}.moveonjoy.com" in line:
            m = FL_RE.search(line)
            if m:
                yield idx, m.group(0), m.group(2)

def enumerate_all_fl_lines(lines):
    """Yield (idx, full_url, relative_path, subdomain) for all moveonjoy lines"""
    for idx, line in enumerate(lines):
        if "http" in line and "moveonjoy.com" in line:
            m = FL_RE.search(line)
            if m:
                yield idx, m.group(0), m.group(2), m.group(1)

def extract_current_subdomain_from_lines(lines):
    """Return the first flNN in the playlist lines, or None."""
    for line in lines:
        m = FL_RE.search(line)
        if m:
            return m.group(1)
    return None

# ---------- LIVENESS & SEARCH ----------

def subdomain_alive_any(lines, subdomain, limit=SAMPLE_LIMIT):
    """
    Consider subdomain alive if ANY channel on that subdomain is playable.
    Sample up to 'limit' channels for speed.
    """
    checked = 0
    for idx, url, path in enumerate_lines_using_subdomain(lines, subdomain):
        checked += 1
        log(f"Testing {url} ...")
        if check_stream_playable(url):
            log(" -> playable (subdomain alive).")
            return True
        else:
            log(" -> not playable.")
        if checked >= limit:
            break
    return False

def find_fallback_for_path(path, exclude_sub=None):
    """Find first flNN (50->3) where the given path is playable."""
    for n in SUB_RANGE:
        sub = f"fl{n}"
        if exclude_sub and sub == exclude_sub:
            continue
        url = f"https://{sub}.moveonjoy.com/{path}"
        log("Trying", url)
        if check_stream_playable(url):
            log(" -> path works on", sub)
            return sub
    return None

def find_any_fallback_main(lines, exclude_sub=None):
    """Find first subdomain where at least one channel is playable (50->3)."""
    for n in SUB_RANGE:
        sub = f"fl{n}"
        if exclude_sub and sub == exclude_sub:
            continue
        log("Probing subdomain", sub)
        if subdomain_alive_any(lines, sub, limit=SAMPLE_LIMIT):
            return sub
    return None

# ---------- PER-CHANNEL FAILOVER & AUTO-RESTORE ----------

def per_channel_failover(lines, current_main):
    """
    For each line on current_main, if that path is not playable on current_main,
    switch that line to a working flNN for that path.
    Returns (new_lines, changed_flag)
    """
    new_lines = list(lines)
    changed = False

    for idx, full_url, path in enumerate_lines_using_subdomain(lines, current_main):
        # path may already be on non-main (skip)
        # test path on main
        test_url = f"https://{current_main}.moveonjoy.com/{path}"
        log("Checking path on main:", test_url)
        if check_stream_playable(test_url):
            log(" -> works on main; keep it.")
            continue

        # Path failing on main; search per-path fallback
        fallback = find_fallback_for_path(path, exclude_sub=current_main)
        if fallback:
            new_url = f"https://{fallback}.moveonjoy.com/{path}"
            log(f"Switching line index {idx} path {path} to fallback {fallback}")
            new_lines[idx] = re.sub(r"https://fl\d+\.moveonjoy\.com/[^\s]+", new_url, new_lines[idx])
            changed = True
        else:
            log(f"No fallback found for path {path}; leaving as-is")

    return new_lines, changed

def auto_restore_to_main(lines, current_main):
    """
    For lines that currently point to other flNN, if current_main now serves the same path,
    restore those lines back to current_main.
    Returns (new_lines, changed_flag)
    """
    new_lines = list(lines)
    changed = False
    for idx, full_url, path, sub in enumerate_all_fl_lines(lines):
        if sub == current_main:
            continue
        main_test = f"https://{current_main}.moveonjoy.com/{path}"
        log("Checking if main now serves", path)
        if check_stream_playable(main_test):
            log(f"Restoring {path} to main {current_main}")
            new_lines[idx] = re.sub(r"https://fl\d+\.moveonjoy\.com/[^\s]+", main_test, new_lines[idx])
            changed = True
    return new_lines, changed

# ---------- GIT & WRITE ----------

def git_commit_and_push_if_changed(m3u_path: Path, old_text: str, new_lines, cooldown=COOLDOWN_SECONDS):
    new_text = "\n".join(new_lines)
    if new_text == old_text:
        log("No textual changes to playlist — nothing to commit.")
        return False

    # cooldown check
    now = int(time.time())
    last = 0
    if LAST_UPDATE_FILE.exists():
        try:
            last = int(LAST_UPDATE_FILE.read_text())
        except Exception:
            last = 0

    # write file regardless (so the runner and any downstream steps see updated playlist)
    m3u_path.write_text(new_text + ("\n" if not new_text.endswith("\n") else ""), encoding="utf-8")
    log("Playlist file updated on disk.")

    if cooldown and (now - last) < cooldown:
        log(f"Cooldown active ({now-last}s since last push) — skipping git push.")
        return True

    # configure git (best-effort)
    try:
        subprocess.run(["git", "config", "--global", "user.email", "actions@github.com"], check=True)
        subprocess.run(["git", "config", "--global", "user.name", "github-actions"], check=True)
    except Exception as e:
        log("Git config warning:", e)

    try:
        subprocess.run(["git", "add", str(m3u_path)], check=True)
        res = subprocess.run(["git", "commit", "-m", f"Auto-update MoveOnJoy subdomains at {datetime.utcnow().isoformat()}"], check=False)
        if res.returncode == 0:
            subprocess.run(["git", "push"], check=True)
            LAST_UPDATE_FILE.write_text(str(now))
            log("Git push completed.")
        else:
            log("Nothing to commit (git returned no changes).")
    except subprocess.CalledProcessError as e:
        log("Git operation failed:", e)
    except Exception as e:
        log("Unexpected git error:", e)

    return True

# ---------- MAIN FLOW ----------

def main():
    log("MoveOnJoy Updater v4.1 starting")
    # debug working dir + list top-level files (helps GH Actions)
    log("Working dir:", Path(".").resolve())
    log("Top-level entries:")
    for itm in sorted(Path(".").iterdir()):
        log(" -", itm)

    p = find_playlist()
    if not p:
        log("ERROR: us.m3u not found. Ensure playlist exists at PrimeVision/us.m3u or run from repo root.")
        sys.exit(1)

    lines, old_text = read_playlist_lines(p)
    current_main = extract_current_subdomain_from_lines(lines)
    if not current_main:
        log("ERROR: no flNN moveonjoy domain found in playlist.")
        sys.exit(1)

    log("Detected main subdomain:", current_main)

    # 1) If any channel on current_main is playable -> main considered alive
    main_alive = subdomain_alive_any(lines, current_main, limit=SAMPLE_LIMIT)

    if main_alive:
        log(f"Main {current_main} appears alive (at least one playable channel).")
        # 1a) Try to restore channels (on other flXX) back to main if their path works on main
        restored_lines, restored_changed = auto_restore_to_main(lines, current_main)
        if restored_changed:
            log("Restored channels back to main; committing (respecting cooldown).")
            git_commit_and_push_if_changed(p, old_text, restored_lines)
            return

        # 1b) Check channels on main that are individually offline -> per-channel fallback
        repaired_lines, repaired_changed = per_channel_failover(lines, current_main)
        if repaired_changed:
            log("Applied per-channel fallbacks for some main channels; committing (respecting cooldown).")
            git_commit_and_push_if_changed(p, old_text, repaired_lines)
            return

        log("All good — no changes needed.")
        return

    # 2) main appears dead (no playable channel found) -> find a fallback main
    log(f"Main {current_main} appears offline (no playable channels). Searching fallback main...")
    fallback_main = find_any_fallback_main(lines, exclude_sub=current_main)
    if not fallback_main:
        log("No fallback main found. Attempting per-channel fallbacks (may partially recover channels).")
        repaired_lines, repaired_changed = per_channel_failover(lines, current_main)
        if repaired_changed:
            log("Per-channel fallbacks applied; committing (respecting cooldown).")
            git_commit_and_push_if_changed(p, old_text, repaired_lines)
            return
        log("No changes possible. Exiting.")
        return

    log(f"Found fallback main: {fallback_main}. Migrating main channels where possible.")
    new_lines = list(lines)
    changed = False
    for idx, url, path, sub in enumerate_all_fl_lines(lines):
        if path in SPECIAL_CHANNEL_PATHS:
            continue
        # attempt to move this path to fallback_main only if that path works there
        test_url = f"https://{fallback_main}.moveonjoy.com/{path}"
        log("Testing fallback path:", test_url)
        if check_stream_playable(test_url):
            log(f" -> path works on fallback; switching line idx {idx} to {fallback_main}")
            new_lines[idx] = re.sub(r"https://fl\d+\.moveonjoy\.com/[^\s]+", test_url, new_lines[idx])
            changed = True

    if changed:
        log("Writing playlist updates switching main channels to fallback main.")
        git_commit_and_push_if_changed(p, old_text, new_lines)
    else:
        log("Found fallback main but no matching paths available on it; no changes performed.")

if __name__ == "__main__":
    main()

File: scripts/moveonjoy_updater.js
#!/usr/bin/env node
/**
 * moveonjoy_updater_full.js
 * Full intelligent MoveOnJoy Updater (Node.js)
 */

const fs = require('fs');
const fsSync = require('fs');
const path = require('path');
const axios = require('axios');
const { spawnSync } = require('child_process');

const PLAYLIST_REL = path.join('PrimeVision', 'us.m3u');
const PLAYLIST_PATH = path.resolve(PLAYLIST_REL);
const BACKUP_PATH = PLAYLIST_PATH + '.bak';
const RANK_FILE = path.resolve('.subdomain_health.json');
const DOCS_DIR = path.resolve('docs'); // for GitHub Pages
const DASHBOARD_HTML = path.join(DOCS_DIR, 'index.html');

const SUB_MIN = 3;
const SUB_MAX = 50;
const DEFAULT_SEARCH_ATTEMPTS = 12; // balanced S2 but wider, uses learned order first
const STABLE_TRIES = 3;
const STABLE_DELAY_MS = 700;
const SAMPLE_LIMIT = 3;
const THREADS = 12;
const COOLDOWN_SECONDS = 3600;
const LAST_UPDATE = '.moveonjoy_last_update';

const FL_RE = /https?:\/\/(fl\d+)\.moveonjoy\.com\/([^\s]+)/i;
const SPECIAL_PATHS = new Set(['DISNEY/index.m3u8']);

// helper logging
function log(...a){ console.log(new Date().toISOString(), ...a); }
function sleep(ms){ return new Promise(r=>setTimeout(r,ms)); }

async function findPlaylist(){
  if (fsSync.existsSync(PLAYLIST_PATH)) return PLAYLIST_PATH;
  // recursive search
  const found = (function walk(dir){
    const L = fsSync.readdirSync(dir, { withFileTypes: true });
    for (const e of L){
      const p = path.join(dir, e.name);
      if (e.isFile() && e.name === 'us.m3u') return p;
      if (e.isDirectory()){
        const r = walk(p);
        if (r) return r;
      }
    }
    return null;
  })('.');
  return found;
}

async function axiosHead(url, timeout=3000){
  try {
    return await axios.head(url, { timeout, maxRedirects: 5, validateStatus: null });
  } catch(e){ return null; }
}
async function axiosGetBytes(url, timeout=6000, range='bytes=0-65535'){
  try {
    return await axios.get(url, { timeout, responseType: 'arraybuffer', headers: { Range: range }, maxRedirects: 5, validateStatus: null });
  } catch(e){ return null; }
}

// quick check: HEAD then small GET if needed
async function fastCheckM3U(url){
  const h = await axiosHead(url);
  if (h && h.status >= 200 && h.status < 400){
    const ct = (h.headers['content-type'] || '').toLowerCase();
    if (ct.includes('mpegurl') || ct.includes('vnd.apple.mpegurl')) return true;
    // try small GET
    const g = await axiosGetBytes(url);
    if (g && g.status >= 200 && g.status < 400){
      const txt = Buffer.from(g.data).toString('utf8',0,5000);
      for (const line of txt.split(/\r?\n/)){
        const s = line.trim();
        if (!s || s.startsWith('#')) continue;
        if (s.endsWith('.ts') || s.endsWith('.m3u8')) return true;
      }
    }
  }
  return false;
}

function bufferHasVideoMarkers(buf){
  if (!buf || buf.length < 16) return false;
  const arr = Buffer.from(buf);
  if (arr[0] === 0x47) return true; // ts sync
  const patterns = [ Buffer.from([0,0,0,1,0x67]), Buffer.from([0,0,0,1,0x65]), Buffer.from([0,0,0,1,0x40]) ];
  for(const p of patterns) if (arr.includes(p)) return true;
  return false;
}

async function checkStreamPlayable(url){
  if (! await fastCheckM3U(url)) return false;
  const g = await axiosGetBytes(url);
  if (!g || g.status >= 400) return false;
  const text = Buffer.from(g.data).toString('utf8');
  for (const L of text.split(/\r?\n/)){
    const s = L.trim();
    if (!s || s.startsWith('#')) continue;
    let seg = s;
    if (!/^https?:\/\//i.test(s)) seg = url.replace(/\/[^\/]*$/, '') + '/' + s;
    const segHead = await axiosHead(seg);
    if (segHead && segHead.status >=200 && segHead.status < 400){
      const segGet = await axiosGetBytes(seg);
      if (segGet && segGet.status >=200 && segGet.status < 400){
        return bufferHasVideoMarkers(segGet.data);
      }
    }
    break;
  }
  return false;
}

async function ensureStable(url, tries=STABLE_TRIES, delay=STABLE_DELAY_MS){
  for (let i=1;i<=tries;i++){
    // do a fast HEAD-first check
    const okFast = await fastCheckM3U(url);
    if (!okFast){
      log(`Stability: fast check failed for ${url} on ${i}/${tries}`);
      return false;
    }
    // on final try do deeper playable check
    if (i === tries){
      const okPlay = await checkStreamPlayable(url);
      if (!okPlay){
        log(`Stability: deep playable check failed for ${url}`);
        return false;
      }
    }
    if (i < tries) await sleep(delay);
  }
  return true;
}

// subdomain ranking store: counts successes and failures per flXX
function loadRank(){
  try { return JSON.parse(fsSync.readFileSync(RANK_FILE,'utf8')); } catch(e){ return {}; }
}
function saveRank(rank){ try{ fsSync.writeFileSync(RANK_FILE, JSON.stringify(rank, null, 2),'utf8'); }catch(e){} }
function bumpRank(sub, ok){
  const r = loadRank();
  if (!r[sub]) r[sub] = { success:0, fail:0, last: null };
  if (ok) r[sub].success++;
  else r[sub].fail++;
  r[sub].last = Date.now();
  saveRank(r);
}
function rankedSubs(current=null){
  // build list and sort by success rate then success count then recency. include full fl range default if missing
  const r = loadRank();
  const arr = [];
  for (let n=SUB_MAX; n>=SUB_MIN; n--){
    const s = `fl${n}`;
    const data = r[s] || { success:0, fail:0, last:null };
    const rate = data.success + data.fail === 0 ? 0 : data.success / (data.success + data.fail);
    arr.push({ sub:s, rate, success:data.success, fail:data.fail, last:data.last });
  }
  // sort: higher rate first, then success, then newer
  arr.sort((a,b)=>{
    if (b.rate !== a.rate) return b.rate - a.rate;
    if (b.success !== a.success) return b.success - a.success;
    return (b.last||0) - (a.last||0);
  });
  // if current present, try to keep it near front? We'll return the full sorted list; searching logic will pick top K around current later
  return arr.map(x=>x.sub);
}

function extractFlFromUrl(url){
  const m = FL_RE.exec(url);
  return m ? m[1] : null;
}

function extractPathFromUrl(url){
  const m = FL_RE.exec(url);
  return m ? m[2] : null;
}

function replaceFlInUrl(url, fl){
  return url.replace(/fl\d+/, fl);
}

// balanced S2 search builder: prefer learned order, but attempt up to maxAttempts
function buildBalancedSearchList(currentFl, maxAttempts=DEFAULT_SEARCH_ATTEMPTS){
  const ranked = rankedSubs(currentFl); // fl list high->low by history
  // start with ranked order but put currentFl near middle; remove current from list
  const list = ranked.filter(s => s !== currentFl);
  // cap list length
  return list.slice(0, maxAttempts);
}

// read & write utilities
function readLines(p){ return fsSync.readFileSync(p,'utf8').split(/\r?\n/); }
function writeLines(p, lines){ fsSync.writeFileSync(p, lines.join('\n') + '\n', 'utf8'); }

// main search for path fallback
async function findWorkingSubForPath(path, exclude=null, current=null){
  const candidates = buildBalancedSearchList(current, DEFAULT_SEARCH_ATTEMPTS);
  for (const sub of candidates){
    if (sub === exclude) continue;
    const url = `https://${sub}.moveonjoy.com/${path}`;
    try {
      const ok = await fastCheckM3U(url);
      if (!ok) continue;
      if (await ensureStable(url)){
        bumpRank(sub, true);
        return sub;
      } else {
        bumpRank(sub, false);
      }
    } catch(e){
      bumpRank(sub, false);
    }
  }
  return null;
}

// Per-channel fallback: change only lines referencing currentMain that fail
async function perChannelFailover(lines, currentMain){
  let changed = false;
  const newLines = lines.slice();
  for (let idx=0; idx<lines.length; idx++){
    const line = lines[idx];
    if (!line || !line.includes('moveonjoy.com')) continue;
    const fl = extractFlFromUrl(line);
    if (fl !== currentMain) continue;
    const path = extractPathFromUrl(line);
    if (!path) continue;
    // test path on current main quickly
    const mainUrl = `https://${currentMain}.moveonjoy.com/${path}`;
    const okMain = await fastCheckM3U(mainUrl);
    if (okMain) continue; // fine
    log(`Channel ${path} fails on ${currentMain}. Searching fallback...`);
    const fallback = await findWorkingSubForPath(path, currentMain, currentMain);
    if (fallback){
      const newUrl = `https://${fallback}.moveonjoy.com/${path}`;
      newLines[idx] = newUrl;
      changed = true;
      log(`Replaced ${path} → ${newUrl}`);
    } else {
      // comment out (A2) with reason and timestamp after repeated fails (safety: do only after 2 search attempts)
      log(`No fallback found for ${path}; commenting out line.`);
      newLines[idx] = `# OFFLINE: ${new Date().toISOString()} - ${line}`;
      changed = true;
    }
  }
  return { newLines, changed };
}

// Auto-restore: if currentMain starts serving a path again, restore lines that point to other flXX
async function autoRestore(lines, currentMain){
  let changed = false;
  const newLines = lines.slice();
  // check lines that reference moveonjoy but not currentMain
  for (let idx=0; idx<lines.length; idx++){
    const line = lines[idx];
    if (!line || !line.includes('moveonjoy.com')) continue;
    const fl = extractFlFromUrl(line);
    if (!fl || fl === currentMain) continue;
    const path = extractPathFromUrl(line);
    if (!path) continue;
    const mainTry = `https://${currentMain}.moveonjoy.com/${path}`;
    if (await ensureStable(mainTry)){
      newLines[idx] = mainTry;
      changed = true;
      log(`Restored ${path} back to main ${currentMain}`);
    }
  }
  return { newLines, changed };
}

function gitCommitAndPushIfChanged(filePath, oldText, newLines){
  const newText = newLines.join('\n') + '\n';
  if (newText === oldText) { log('No changes to commit'); return false; }
  try { fsSync.writeFileSync(BACKUP_PATH, oldText, 'utf8'); } catch(e){}
  fsSync.writeFileSync(filePath, newText, 'utf8');
  log('Playlist written to disk and backed up.');

  // cooldown
  const now = Math.floor(Date.now()/1000);
  let last = 0;
  try { if (fsSync.existsSync(LAST_UPDATE)) last = parseInt(fsSync.readFileSync(LAST_UPDATE,'utf8'))||0; } catch(e){ last=0; }
  if (COOLDOWN_SECONDS && (now - last) < COOLDOWN_SECONDS){
    log('Cooldown active — skipping git push.');
    return true;
  }

  try {
    spawnSync('git', ['config', '--global', 'user.email', 'actions@github.com']);
    spawnSync('git', ['config', '--global', 'user.name', 'github-actions']);
    spawnSync('git', ['add', filePath], { stdio:'inherit' });
    const res = spawnSync('git', ['commit', '-m', `Auto-update MoveOnJoy subdomains at ${new Date().toISOString()}`], { stdio:'inherit' });
    if (res.status === 0) {
      spawnSync('git', ['push'], { stdio:'inherit' });
      fsSync.writeFileSync(LAST_UPDATE, String(now),'utf8');
      log('Changes pushed.');
    } else {
      log('Nothing to commit (git returned no change).');
    }
  } catch(e){
    log('Git operation failed:', e);
  }
  return true;
}

// Build a small docs HTML page for dashboard (C1)
function buildDashboardHtml(channelStatuses, rankedList){
  let rows = '';
  for (const s of channelStatuses){
    rows += `<tr><td>${s.title||'?'}</td><td>${s.path}</td><td>${s.url}</td><td>${s.status}</td><td>${s.note||''}</td></tr>\n`;
  }
  const rankedHtml = rankedList.map((r,i)=>`<li>${r}</li>`).join('\n');
  return `<!doctype html>
<html>
<head><meta charset="utf-8"><title>MoveOnJoy Updater Dashboard</title>
<style>body{font-family:system-ui,Arial}table{border-collapse:collapse;width:100%}td,th{border:1px solid #ddd;padding:8px}</style>
</head><body>
<h1>MoveOnJoy Updater Dashboard</h1>
<p>Generated: ${new Date().toISOString()}</p>
<h2>Subdomain ranking (most reliable → )</h2><ol>${rankedHtml}</ol>
<h2>Channels</h2>
<table><thead><tr><th>Title</th><th>Path</th><th>URL</th><th>Status</th><th>Note</th></tr></thead><tbody>${rows}</tbody></table>
</body></html>`;
}

// main run
(async function main(){
  try {
    log('Starting MoveOnJoy Updater FULL');
    const playlist = await findPlaylist();
    if (!playlist) { log('us.m3u not found; abort'); process.exit(1); }
    log('Playlist:', playlist);
    const raw = fsSync.readFileSync(playlist,'utf8');
    const lines = raw.split(/\r?\n/);

    const currentMain = (function findFirstFl(){
      for (const L of lines){
        const m = FL_RE.exec(L);
        if (m) return m[1];
      }
      return null;
    })();
    if (!currentMain){ log('No flNN main found; abort'); process.exit(1); }
    log('Detected current main', currentMain);

    // 1) If main alive by any sample → auto-restore then per-channel fix
    let sampleAlive = false;
    let sampleCount = 0;
    for (const e of lines){
      if (e && e.includes(currentMain) && e.includes('moveonjoy.com')){
        sampleCount++;
        const ok = await fastCheckM3U(e);
        if (ok){ sampleAlive = true; break; }
        if (sampleCount >= SAMPLE_LIMIT) break;
      }
    }

    if (sampleAlive){
      log('Main appears alive; attempting restore & per-channel fix');
      const { newLines: restored, changed: restoredChanged } = await autoRestore(lines, currentMain);
      if (restoredChanged){ gitCommitAndPushIfChanged(playlist, raw, restored); return; }
      const { newLines: fixed, changed: fixedChanged } = await perChannelFailover(lines, currentMain);
      if (fixedChanged){ gitCommitAndPushIfChanged(playlist, raw, fixed); return; }
      log('No changes required.');
    } else {
      log('Main appears offline (sample failed). Searching fallback main');
      // find any fallback main (fast check on ranked subs)
      const ranked = rankedSubs(currentMain);
      let foundMain = null;
      for (const cand of ranked.slice(0, DEFAULT_SEARCH_ATTEMPTS)){
        if (cand === currentMain) continue;
        // test a sample of channels for cand
        let anyOk = false;
        for (const L of lines){
          if (L && L.includes(`${cand}.moveonjoy.com`)){
            if (await fastCheckM3U(L)) { anyOk = true; break; }
          }
        }
        if (anyOk){ foundMain = cand; break; }
      }
      if (!foundMain){
        log('No fallback main found; attempt per-channel fallbacks');
        const { newLines: fixed, changed: fixedChanged } = await perChannelFailover(lines, currentMain);
        if (fixedChanged){ gitCommitAndPushIfChanged(playlist, raw, fixed); return; }
        log('No changes possible.');
      } else {
        log('Found fallback main', foundMain, ' — migrating matching paths');
        const newLines = lines.slice();
        let changed = false;
        for (let idx=0; idx<lines.length; idx++){
          const L = lines[idx];
          if (!L || !L.includes('moveonjoy.com')) continue;
          const pathPart = extractPathFromUrl(L);
          if (!pathPart) continue;
          if (SPECIAL_PATHS.has(pathPart)) continue;
          const testUrl = `https://${foundMain}.moveonjoy.com/${pathPart}`;
          if (await fastCheckM3U(testUrl) && await ensureStable(testUrl)){
            newLines[idx] = testUrl;
            changed = true;
            log('Switched',pathPart,'->',foundMain);
          }
        }
        if (changed) gitCommitAndPushIfChanged(playlist, raw, newLines);
        else log('Fallback main found but no matching stable paths on it.');
      }
    }

    // update dashboard and ranking file
    try {
      const rankList = rankedSubs(currentMain);
      // compute channel statuses for dashboard
      const chs = [];
      for (const L of lines){
        if (!L) continue;
        let status='N/A', note='';
        if (L.includes('moveonjoy.com')) {
          const fl = extractFlFromUrl(L);
          const pth = extractPathFromUrl(L);
          const playable = await fastCheckM3U(L);
          status = playable ? 'ONLINE' : 'OFFLINE';
          chs.push({ title:'', path: pth||'', url: L, status, note });
        }
      }
      if (!fsSync.existsSync(DOCS_DIR)) fsSync.mkdirSync(DOCS_DIR);
      fsSync.writeFileSync(DASHBOARD_HTML, buildDashboardHtml(chs, rankList), 'utf8');
      // commit docs
      spawnSync('git', ['add', DASHBOARD_HTML]);
      spawnSync('git', ['commit', '-m', `Update MoveOnJoy dashboard ${new Date().toISOString()}`], { stdio:'inherit' });
      spawnSync('git', ['push'], { stdio:'inherit' });
    } catch(e){ log('Dashboard update failed:', e); }

    log('Done run.');
  } catch(e){ log('Fatal error:', e); process.exit(1); }
})();